scripts/cdc-sabeti-covid-19/batch_upsert_entities.py

```diff
@@ -0,0 +1,138 @@
+                        "gisaid_fasta", "gisaid_meta_tsv", "ivar_trim_stats_html", "ivar_trim_stats_png",
+                        "ivar_trim_stats_tsv", "max_ntc_bases", "meta_by_filename_json",
+                        "multiqc_report_cleaned", "multiqc_report_raw", "nextclade_all_json", "nextclade_auspice_json",
+                        "nextmeta_tsv", "num_assembled", "num_failed_annotation", "num_failed_assembly",
+                        "num_read_files", "num_samples", "num_submittable", "picard_metrics_wgs",
+                        "run_date", "sequencing_reports", "spikein_counts", "sra_metadata", "submission_xml",
+                        "submission_zip", "submit_ready"]
+
+    # templates for request body components
+    template_req_body = '''[{"name":"VAR_ENTITY_ID",
+                             "entityType":"VAR_ENTITY_TYPE",
+                             "operations":[OPERATIONS_LIST]}]'''
+
+    template_make_list_attr = '{"op":"CreateAttributeValueList","attributeName":"VAR_ATTRIBUTE_LIST_NAME"},'
+    template_add_list_member = '{"op":"AddListMember","attributeListName":"VAR_ATTRIBUTE_LIST_NAME", "newMember":"VAR_LIST_MEMBER"},'
+    template_make_single_attr = '{"op":"AddUpdateAttribute","attributeName":"VAR_ATTRIBUTE_NAME", "addUpdateAttribute":"VAR_ATTRIBUTE_MEMBER"},'
+
+    # initiate string to capture all operation requests
+    all_operation_requests = ''''''
+
+    # if there are single attribute columns
+    if single_attr_cols:
+        # for each column that is not an array
+        for col in single_attr_cols:
+            # get valuue in col from df
+            attr_val = str(df_tsv.iloc[0][col])
+
+            # add the request for attribute to list
+            all_operation_requests += template_make_single_attr.replace("VAR_ATTRIBUTE_MEMBER", attr_val).replace("VAR_ATTRIBUTE_NAME", col)
+
+    # if there are array attribute columns
+    if array_attr_cols:
+        # for each column that is an array
+        for col in array_attr_cols:
+            all_operation_requests += template_make_list_attr.replace("VAR_ATTRIBUTE_LIST_NAME", col)
+            # convert "array" from tsv which translates to a string back into an array
+            attr_values = str(df_tsv.iloc[0][col]).replace('"', '').strip('[]').split(",")
+            for val in attr_values:
+                all_operation_requests += template_add_list_member.replace("VAR_LIST_MEMBER", val).replace("VAR_ATTRIBUTE_LIST_NAME", col)
+
+    # remove trailing comma from the last request template
+    all_operation_requests = all_operation_requests[:-1]
+
+    # get the entity_type (table name) and entity_id (row id - must be unique)
+    entity_id = str(df_tsv.iloc[0][0])
+    entity_type = entity_type_col_name.rsplit("_", 1)[0].split(":")[1]
+
+    # put entity_type and entity_id in request body template
+    final_request = template_req_body.replace("VAR_ENTITY_ID", entity_id).replace("VAR_ENTITY_TYPE", entity_type)
```

Comment:
It might be better to give each replace a new varable

Reply:
I did have it separated out and then read that you can combine replace statements together! I thought it would be more concise and also tested it thoroughly to make sure that it doesn't do anything unwanted.